### PR TITLE
Support simple limiter in memory

### PR DESCRIPTION
**Pull request by ponytailer**
- support the limiter in memory
- fix some typo in tests

### CODE REVIEW

File: ratelimit/backends/simple.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    def remove_rule_later(self, path: str, rule_key: str):
+        _, deadline = self.blocks[path][rule_key]
+        return self.call_at(deadline, self.remove_rule, path, rule_key)
+
+    def set_blocked_user(self, user: str, block_time: int) -> int:
+        self.blocked_users[user] = block_time + self.now()
+        self.remove_blocked_user_later(user)
+        return block_time
+
+    def set_rule(self, rules: Dict, path: str, rule: str, limit: int, timestamp: int):
```
Comment (abersheeran) [inline ratelimit/backends/simple.py:67]: Is it `Dict[str, Tuple[int, int]]`?

File: ratelimit/backends/simple.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        rules[rule] = [limit - 1, timestamp]
```
Comment (abersheeran) [inline ratelimit/backends/simple.py:68]: Tuple better than List at here.

File: ratelimit/backends/simple.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    @synchronized
+    def remove_rule(self, path: str, rule_key: str) -> Optional[List[int]]:
+        return self.blocks[path].pop(rule_key, None)
+
+    def remove_blocked_user_later(self, user: str):
+        later = self.blocked_users[user]
+        self.call_at(later, self.remove_user, user)
+
+    def remove_rule_later(self, path: str, rule_key: str):
```
Comment (abersheeran) [inline ratelimit/backends/simple.py:58]: Missing return type.

File: ratelimit/backends/simple.py (hunk 2 above)
Comment (ponytailer) [reply ratelimit/backends/simple.py:68]: Yes， I know。But the limit in tuple can not be change。

Comment (abersheeran) [reply ratelimit/backends/simple.py:68]: I see.

File: ratelimit/backends/simple.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+from threading import Lock
+from typing import Dict, List, Optional
+
+from ..rule import Rule
+from . import BaseBackend
+
+lock = Lock()
+
+
+def synchronized(func):
```
Comment (abersheeran) [inline ratelimit/backends/simple.py:13]: I prefer to put `Lock` in the instance, one storage structure corresponds to one lock.

File: ratelimit/backends/simple.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    def __init__(self):
+        # user: deadline
+        self.blocked_users: Dict[str, int] = {}
+        # path: {rule_key: (limit, timestamp)}
+        self.blocks: Dict[str, Dict[str, List[int]]] = defaultdict(dict)
+
+    @staticmethod
+    def now() -> int:
+        return int(time.time())
```
Comment (abersheeran) [inline ratelimit/backends/simple.py:35]: https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.time is better in the case.

File: ratelimit/backends/simple.py (hunk 4 above)
Comment (ponytailer) [reply ratelimit/backends/simple.py:10]: OK

File: pyproject.toml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "asgi-ratelimit"
-version = "0.6.1"
+version = "0.6.2"
```
Comment (abersheeran) [inline pyproject.toml:3]: We should use 0.7.0 for the feature.

**Review by abersheeran — CHANGES_REQUESTED** [review]

File: ratelimit/backends/simple.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class MemoryBackend(BaseBackend):
+    """simple limiter with memory"""
+
+    def __init__(self):
+        # user: deadline
+        self.blocked_users: Dict[str, int] = {}
+        # path: {rule_key: (limit, timestamp)}
+        self.blocks: Dict[str, Dict[str, List[int]]] = defaultdict(dict)
```
Comment (abersheeran) [inline ratelimit/backends/simple.py:29]: Maybe I didn't express it clearly, I mean that each pair of blocked_users and blocks corresponds to a `Lock` 😂 We should lock the multi-threaded concurrent operations on the data structure, right?

**Review by abersheeran — CHANGES_REQUESTED** [review]

File: ratelimit/backends/simple.py (hunk 7 above)
Comment (ponytailer) [reply ratelimit/backends/simple.py:29]: My mistake. But the `dict` is thread-safe, the `lock` is unnecessary maybe ?

Comment (abersheeran) [reply ratelimit/backends/simple.py:29]: > My mistake. But the dict is thread-safe, the lock is unnecessary maybe ?

`self.blocks[path].pop(rule_key, None)` is not thread-safe. 

---

`retry_after` need a Lock to keep number decr is thread-safe

**Review by abersheeran — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**